tests/support/helpers.py

```diff
@@ -1067,6 +1067,111 @@ def decorator(*args, **kwargs):
     return decorator
 
 
+@requires_system_grains
+def runs_on(grains=None, **kwargs):
+    '''
+    Skip the test if grains don't match the values passed into **kwargs
+    if a kwarg value is a list then skip if the grains don't match any item in the list
+    '''
+    def decorator(caller):
+        @functools.wraps(caller)
+        def wrapper(cls):
+            for kw, value in kwargs.items():
+                if isinstance(value, list):
+                    if not any(str(grains.get(kw)).lower() != str(v).lower() for v in value):
+                        cls.skipTest('This test does not run on {}={}'.format(kw, grains.get(kw)))
+                else:
+                    if str(grains.get(kw)).lower() != str(value).lower():
+                        cls.skipTest('This test runs on {}={}, not {}'.format(kw, value, grains.get(kw)))
+            return caller(cls)
+        return wrapper
+    return decorator
+
+
+@requires_system_grains
+def not_runs_on(grains=None, **kwargs):
+    '''
+    Reverse of `runs_on`.
+    Skip the test if any grains match the values passed into **kwargs
+    if a kwarg value is a list then skip if the grains match any item in the list
+    '''
+    def decorator(caller):
+        @functools.wraps(caller)
+        def wrapper(cls):
+            for kw, value in kwargs.items():
+                if isinstance(value, list):
+                    if any(str(grains.get(kw)).lower() == str(v).lower() for v in value):
+                        cls.skipTest('This test does not run on {}={}'.format(kw, grains.get(kw)))
+                else:
+                    if str(grains.get(kw)).lower() == str(value).lower():
+                        cls.skipTest('This test does not run on {}={}, got {}'.format(kw, value, grains.get(kw)))
+            return caller(cls)
+        return wrapper
+    return decorator
+
+
+def requires_salt_state_modules(*names):
+    '''
+    Makes sure the passed salt state module is available. Skips the test if not
+
+    .. versionadded:: 3000
+    '''
+
+    def _check_required_salt_state_modules(*required_salt_modules):
```

Comment:
Does this have to be a nested helper function? I'd prefer something like this live at the top level of the module.

Reply:
Was just following the format of `requires_salt_modules`, they share quite a bit of code though and it certainly could sit at the top level